apps/studio/src/lib/db/clients/BasicDatabaseClient.ts

```diff
@@ -379,14 +381,13 @@ export abstract class BasicDatabaseClient<RawResultType extends BaseQueryResult>
     throw new Error("Not implemented");
   }
 
-  async getInsertQuery(tableInsert: TableInsert): Promise<string> {
+  protected createUpsertFunc: ((table: DatabaseEntity, data: {[key: string]: any}, primaryKey: string[]) => string) | null = null
+
+  async getInsertQuery(tableInsert: TableInsert, runAsUpsert = false): Promise<string> {
     const columns = await this.listTableColumns(tableInsert.table, tableInsert.schema);
-    tableInsert.data.forEach((row) => {
-      Object.keys(row).forEach((key) => {
-        row[key] = this.deserializeValue(row[key]);
-      })
-    })
-    return buildInsertQuery(this.knex, tableInsert, columns);
+    const primaryKeysPromise = await this.getPrimaryKeys(tableInsert.table, tableInsert.schema)
+    const primaryKeys = primaryKeysPromise.map(v => v.columnName)
+    return buildInsertQuery(this.knex, tableInsert, { columns, runAsUpsert, primaryKeys, createUpsertFunc: this.createUpsertFunc });
```

Comment:
It looks like we are missing a deserialization. `tableInsert` needs to be deserialized so knex can generate the right query (especially with buffer types). I missed adding the query test for buffers but you can also test this from the table view by right-clicking a row -> `copy as SQL`. This is a sample SQLite table with buffers:

```sql
CREATE TABLE test_table (id INTEGER PRIMARY KEY AUTOINCREMENT, data BLOB NOT NULL);
INSERT INTO test_table (data) VALUES (X'DEADBEEF'), (X'CAFEBABE');
```

P.S. I hate that the deserialization does nested loop like that :). There must be a better way probably.

Reply:
@azmy60 , looks like a snippet got lost probably in one of the million merge conflicts that happened since starting this :D I added the snippet back and got the expected behavior.

Reply:
😆 thanks!